File: least_squares_robust.cpp
#include <opencv2/opencv.hpp>
#include <vector>

using namespace cv;
using namespace std;

// Cauchy weight function
double CauchyWeight(double residual, double c) {
    return 1.0 / (1.0 + (residual / c) * (residual / c));
}

// Function to fit a quadratic polynomial surface to the background using Cauchy weights
void fitQuadraticSurfaceRobust(const Mat& src, Mat& dst, double c) {
    // Collect data points
    vector<Point3f> points;
    for (int y = 0; y < src.rows; ++y) {
        for (int x = 0; x < src.cols; ++x) {
            points.push_back(Point3f((float)x, (float)y, src.at<uchar>(y, x)));
        }
    }
    int npts = (int)points.size();

    // Initialize matrices
    Mat A(npts, 6, CV_32F);
    Mat B(npts, 1, CV_32F);
    for (int i = 0; i < npts; ++i) {
        float x = points[i].x;
        float y = points[i].y;
        float z = points[i].z;

        A.at<float>(i, 0) = x * x;
        A.at<float>(i, 1) = y * y;
        A.at<float>(i, 2) = x * y;
        A.at<float>(i, 3) = x;
        A.at<float>(i, 4) = y;
        A.at<float>(i, 5) = 1;
        B.at<float>(i, 0) = z;
    }

    // Initial guess for the coefficients (least squares)
    Mat coeffs;
    solve(A, B, coeffs, DECOMP_SVD);

    // Iterative reweighted least squares
    Mat W = Mat::eye(npts, npts, CV_32F);
    for (int iter = 0; iter < 10; ++iter) {  // 10 iterations for example
        // Calculate residuals
        Mat residuals = A * coeffs - B;

        // Calculate weights
        for (int i = 0; i < npts; i++) {
            W.at<float>(i, i) = (float)CauchyWeight(residuals.at<float>(i, 0), c);
        }

        // Weighted least squares
        Mat Aw = A.t() * W * A;
        Mat Bw = A.t() * W * B;
        solve(Aw, Bw, coeffs, DECOMP_SVD);
    }

    // Create the background model
    Mat bkg = A * coeffs;
    dst = bkg.reshape(0, src.rows);
}

int main(int argc, char** argv) {
    // Load the input image
    Mat image = imread("sample_circle.png", IMREAD_GRAYSCALE);
    if (image.empty()) {
        cout << "Error: Could not open or find the image!" << endl;
        return -1;
    }

    // Fit and remove the background
    Mat background, imagef, residual, binary;
    double c = 2.3849; // Cauchy function parameter
    fitQuadraticSurfaceRobust(image, background, c);
    image.convertTo(imagef, CV_32F);
    subtract(background, imagef, residual);
    background.convertTo(background, CV_8U);
    residual.convertTo(residual, CV_8U);
    threshold(residual, binary, 0, 255, THRESH_BINARY | THRESH_OTSU);

    // Display results
    imshow("Original Image", image);
    imshow("Background", background);
    imshow("Residual", residual);
    imshow("Binary Image", binary);

    waitKey(0);
    return 0;
}

File: least_squares_robust.py
import cv2
import numpy as np
import matplotlib.pyplot as plt
import os
gscale = 2 if "ANDROID_STORAGE" in os.environ else 1
plt.rcParams.update({'font.size': 14*gscale})

def cauchy_weight(residual, c):
    return 1.0 / (1.0 + (residual / c)**2)

def fit_quadratic_surface_robust(image, c, iterations=20):
    rows, cols = image.shape
    X, Y = np.meshgrid(np.arange(cols), np.arange(rows))
    Z = image

    # Prepare the data for least squares fitting
    X_flat = X.flatten()
    Y_flat = Y.flatten()
    Z_flat = Z.flatten()

    A = np.vstack([X_flat**2, Y_flat**2, X_flat*Y_flat, X_flat, Y_flat, np.ones_like(X_flat)]).T
    B = Z_flat    

    # Initial guess for the coefficients (least squares)
    coeffs, _, _, _ = np.linalg.lstsq(A, B, rcond=None)

    # Iterative reweighted least squares
    for itr in range(iterations):
        residuals = A @ coeffs - B
        
        weights = cauchy_weight(residuals, c)
        W = np.diag(weights)

        Aw = A.T @ W @ A
        Bw = A.T @ W @ Z_flat
        coeffs = np.linalg.solve(Aw, Bw)

        # Create the background model
        background = (coeffs[0] * X**2 + coeffs[1] * Y**2 +
                      coeffs[2] * X * Y + coeffs[3] * X +
                      coeffs[4] * Y + coeffs[5])
                      
        residual = background.astype(np.float32) - image.astype(np.float32)
        residual = np.clip(residual, 0, 255).astype(np.uint8)
        _, binarized = cv2.threshold(residual, 0, 255, cv2.THRESH_BINARY | cv2.THRESH_OTSU)                      
                      
        # Plot the intermediate result
        plt.clf()
        plt.subplot(2, 2, 1)
        plt.title(f'Iteration: #{itr + 1}')
        plt.imshow(image, cmap='gray')
        plt.axis('off')
        plt.subplot(2, 2, 2)
        plt.title('Background')
        plt.imshow(background, cmap='gray')
        plt.axis('off')
        plt.subplot(2, 2, 3)
        plt.title('Residual Image')
        plt.imshow(residual, cmap='gray')
        plt.axis('off')
        plt.subplot(2, 2, 4)
        plt.title('Binary Image')
        plt.imshow(binarized, cmap='gray')
        plt.axis('off')
        plt.pause(0.1)

    return background

def main(image_path):
    # Load the input image
    image = cv2.imread(image_path, cv2.IMREAD_GRAYSCALE)
    if image is None:
        print("Error: Could not open or find the image!")
        return

    # Initialize display
    plt.figure(figsize=(7*gscale, 7*gscale))
    plt.pause(0.1)

    # Fit and remove the background
    c = 2.3849  # Cauchy function parameter
    background = fit_quadratic_surface_robust(image, c)
    residual = background.astype(np.float32) - image.astype(np.float32)
    residual = np.clip(residual, 0, 255).astype(np.uint8)
    _, binarized = cv2.threshold(residual, 0, 255, cv2.THRESH_BINARY | cv2.THRESH_OTSU)

    # Display final results
    plt.subplot(2, 2, 1)
    plt.title('Original Image')
    plt.imshow(image, cmap='gray')
    plt.axis('off')
    plt.subplot(2, 2, 2)
    plt.title('Background')
    plt.imshow(background, cmap='gray')
    plt.axis('off')
    plt.subplot(2, 2, 3)
    plt.title('Residual Image')
    plt.imshow(residual, cmap='gray')
    plt.axis('off')
    plt.subplot(2, 2, 4)
    plt.title('Binary Image')
    plt.imshow(binarized, cmap='gray')
    plt.axis('off')
    plt.show()

if __name__ == "__main__":
    main('sample_circle.png')